File: old/E_Tree_Game.py
import sys, math, random; from heapq import heapify, heappop, heappush;
from collections import defaultdict, Counter, deque; from bisect import bisect_left, bisect_right
def IL(): return list(map(int, sys.stdin.readline().strip().split()))
def S(): return sys.stdin.readline().strip()

def solve():
    n = IL()[0]
    adj = [[] for _ in range(n + 1)]
    for _ in range(n - 1):
        u, v = IL()
        adj[u].append(v)
        adj[v].append(u)
    arr = set()

    for start in range(1, n + 1):
        move = 0
        que = deque([start])
        vis = [0] * (n + 1)
        vis[start] = 1
        while que:
            for _ in range(len(que)):
                node = que.popleft()
                if (move & 1 and move > 1):
                    arr.add((min(node, start), max(node, start)))
                
                for nei in adj[node]:
                    if vis[nei] == 0:
                        que.append(nei)
                        vis[nei] = 1
            move += 1


    step = 1
    if len(arr) & 1:
        print("First", flush=True)
    else:
        print("Second", flush=True)
        step = 0

    while True:
        if step & 1:
            i, j = arr.pop()
            print(i, j, flush=True)
        else:
            i, j = IL()
            if i == -1 and j == -1:
                return
            arr.discard((i, j))

        step += 1



tt = 1
# tt = IL()[0]
for _ in range(tt):
    solve()

File: old/E_Tree_Game.cpp
#include "bits/stdc++.h"
#ifndef ONLINE_JUDGE
#include "debug/debug.h"
#endif
using namespace std;
// #define endl '\n'
typedef long long ll;
const int MOD = 1e9 + 7;

void solve() {
    int n; cin >> n;
    vector<vector<int>> adj(n + 1);
    for (int i = 0; i < n - 1; i++) {
        int u, v; cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    unordered_set<vector<int>> arr;
    for (int start = 1; start <= n; start++) {
        int move = 0;
        queue<int> q; q.push(start);
        vector<int> vis(n); vis[start] = 1;

        while (!q.empty()) {
            int sz = q.size();
            for (int i = 0; i < sz; i++) {
                int node = q.front(); q.pop();
                if (move & 1 && move > 1) {
                   
                }
            }
            move++;
        }
    }

    int step = 1;
    if (arr.size() & 1) {
        cout << "First" << endl;
    }else {
        cout << "Second" << endl;
        step = 0;
    }

    while (true) {
        if (step & 1) {
            arr.
        } else {

        }

        step++;
    }



   
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    int t = 1;
    // cin >> t;
    
    while (t--) {
        solve();
    }
    
    return 0;
}